platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,447 @@
+ * All fields are REQUIRED.
+ */
+message RosterEntry {
+
+    /**
+     * A consensus node identifier.
+     * <p>
+     * Node identifiers SHALL be unique _within_ a shard and realm,
+     * but a node SHALL NOT, ever, serve multiple shards or realms,
+     * therefore the node identifier MAY be repeated _between_ shards and realms.
+     */
+    uint64 node_id = 1;
+
+    /**
+     * A consensus weight.
+     * <p>
+     * Each node SHALL have a weight of zero or more in consensus calculations.<br/>
+     * The consensus weight of a node SHALL be calculated based on the amount
+     * of HBAR staked to that node.<br/>
+     * Consensus SHALL be calculated based on agreement of greater than `2/3`
+     * of the total `weight` value of all nodes on the network.
+     */
+    uint64 weight = 2;
+
+    /**
+     * An RSA public certificate used for signing gossip events.
+     * <p>
+     * This value SHALL be a certificate of a type permitted for gossip
+     * signatures.<br/>
+     * This value SHALL be the DER encoding of the certificate presented.<br/>
+     * This field is REQUIRED and MUST NOT be empty.
+     */
+    bytes gossip_ca_certificate = 3;
+
+    /**
+     * An elliptic curve public encryption key.<br/>
+     * This contains the _long term_ public key for this node.
+     * <p>
+     * This value SHALL be the DER encoding of the presented elliptic curve
+     * public key.<br/>
+     * This field is REQUIRED and MUST NOT be empty.
+     */
+    bytes tss_encryption_key = 4;
+
+    /**
+     * A list of service endpoints for gossip.
+     * <p>
+     * These endpoints SHALL represent the published endpoints to which other
+     * consensus nodes may _gossip_ transactions.<br/>
+     * If the network configuration value `gossipFqdnRestricted` is set, then
+     * all endpoints in this list SHALL supply only IP address.<br/>
+     * If the network configuration value `gossipFqdnRestricted` is _not_ set,
+     * then endpoints in this list MAY supply either IP address or FQDN, but
+     * SHALL NOT supply both values for the same endpoint.<br/>
+     * This list SHALL NOT be empty.<br/>
+     * This list SHALL NOT contain more than `10` entries.<br/>
+     * The first two entries in this list SHALL be the endpoints published to
+     * all consensus nodes.<br/>
+     * All other entries SHALL be reserved for future use.
+     */
+    repeated proto.ServiceEndpoint gossip_endpoint = 5;
+}
+```
+
+The map of rosters will be stored in the state as a virtual map of Key `Hash<Roster>` and value `Roster`.
+The `Hash` object(SHA_384), which is the key of the virtual map, will reuse an existing protobuf that looks as follows:
+```proto
+/**
+ * List of hash algorithms
+ */
+enum HashAlgorithm {
+  HASH_ALGORITHM_UNKNOWN = 0;
+  SHA_384 = 1;
+}
+
+/**
+ * Encapsulates an object hash so that additional hash algorithms
+ * can be added in the future without requiring a breaking change.
+ */
+message HashObject {
```

Comment:
It's intended to be the key of the virtual map, so it must be a protobuf object. And since the Hash protobuf object already exist, we reuse it. I've only included it here for completeness.

Reply:
Where does it already exist?   Can you show me pre-existing usage?  I'd like to learn.

Reply:
We shouldn't reuse the "streams" HashObject and HashAlgorithm.
Both are not well structured (something Jasper brought up in a different context).
The Block Stream is adding a better defined HashAlgorithm to basic_types.proto:
```protobuf
/**
 * A specific hash algorithm.
 *
 * We did not reuse Record Stream `HashAlgorithm` here because in all cases,
 * currently, this will be `SHA_384` and if that is the default value then
 * we can save space by not serializing it, whereas `HASH_ALGORITHM_UNKNOWN`
 * is the default for Record Stream `HashAlgorithm`.
 */
enum BlockHashAlgorithm {
    /**
     * A SHA2 algorithm SHA-384 hash.
     * <p>
     * This is the default value, if a field of this enumerated type is
     * not set, then this is the value that will be decoded when the
     * serialized message is read.
     */
    SHA_384 = 0;
}
```

Regarding `HashObject`, it might be better to just use a `bytes` for the hash value directly, and specify the hash algorithm, where needed (which I believe is somewhat rare), on the message that contains the hash value.

Reply:
thanks @jsync-swirlds for the context. @edward-swirldslabs I considered the HashObject as I noticed it's used in a bunch of utils(`RecordSigningUtils` and `FileSignTool` are examples). Additionally, the initial map implementation was supposed to be PBJ Map, which does not allow byte-type keys. 

We have however since concluded on using `VirtualMap` instead. So, to use a `byte` key as suggested, we will have to provide an implementation of `VirtualKey` or similar. I've removed the HashObject and replaced with this explanation.

Reply:
@kfa-aguda, when you say that the VirtualMap doesn't allow bytes for keys, are you saying this is not possible as a key? 

```
message WrappedKey{
   bytes byteArray;
}

```

Reply:
I looked it up, only integers and strings are allowed for keys.  complex objects are not allowed as keys in pbj maps.    This issue is a non-issue since we're not using pbj maps here.

Reply:
This probably works the PBJ map. I didn't implement it as subsequent conversations shifted to using `VirtualMap`. Now, we're back to using PBJ map, so I've consulted @imalygin for recommendation. I think a simple byte wrapper like you proposed should be sufficient.

Reply:
If we're wrapping bytes, it may be appropriate to use the "well known type" `google.protobuf.BytesValue` rather than creating our own message for the same purpose.

Reply:
agreed on using a well known type, although @ivan suggested `com.hedera.pbj.runtime.io.buffer.Bytes`. Happy to work that out during implementation.